executor/runcexecutor/executor.go

```diff
@@ -524,14 +532,50 @@ func (p *procHandle) WaitForReady(ctx context.Context) error {
 func (p *procHandle) WaitForStart(ctx context.Context, startedCh <-chan int, started func()) error {
 	startedCtx, timeout := context.WithTimeout(ctx, 10*time.Second)
 	defer timeout()
-	var err error
 	select {
 	case <-startedCtx.Done():
 		return errors.New("runc started message never received")
-	case pid, ok := <-startedCh:
+	case runcPid, ok := <-startedCh:
 		if !ok {
 			return errors.New("runc process failed to send pid")
 		}
+
+		// we need the process for the runc monitor process so we can send
+		// SIGWINCH signals to it, if we send them to the process inside the
+		// container nothing happens.
+		var err error
+		p.MonitorProcess, err = os.FindProcess(runcPid)
+		if err != nil {
+			return errors.Wrapf(err, "unable to find runc monitor process for pid %d", runcPid)
+		}
+
+		// the pid reported from the started channel seems to be the pid
+		// of the runc monitor process, but we will need to send signals
+		// to the process running in the container so load the process pid
+		// from the pidfile.  Unfortunately the pidfile will not be written
+		// by the time we get the runc pid via the channel, so we need to retry.
+		var pid int
+		for {
+			pidData, err := os.ReadFile(p.pidfile)
+			if err != nil {
+				return errors.Wrap(err, "unable to read pidfile after runc process started")
+			}
+			pid, err = strconv.Atoi(string(pidData))
```

Comment:
In theory, runc could partially write the pidfile while we are reading it?  Not sure if this is actually possible.  Also not sure how we could know, perhaps we could just add the os.FindProcess call fto the retry loop?

Reply:
1. runc creates pidfile atomically (create + rename) so no partial read is possible.
2. AFAIK os.FIndProcess() is a no-op on Linux (it merely wraps a pid in a structure and returns it). Meaning, it doesn't actually check whether the process exists, it never returns any errors. This might be changed in the future of course.